### PR TITLE
☝🏼Add new prefix field in connection page (prepended to the stream name) 

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
Introduce new prefix field on the sync page to use as prefix for destination tables

Implements #2239
(see #2335 too)

## How
*Describe the solution*
I introduced a new `airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/AirbyteMapper.java` hierarchy class to mimic the same as `Source` and `Destination`

This Is passed to the `DefaultSyncWorker` constructor and does only one transformation about the namespace field for the moment.
It mutates the catalog object config before passing it to the AirbyteDestination but doesn't affect the catalog from the source

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java`
1. `airbyte-api/src/main/openapi/config.yaml`
1. the rest

### CODE REVIEW

File: airbyte-api/src/main/openapi/config.yaml [hunk 1]
```diff
@@ -1595,13 +1595,17 @@ components:
...
       required:
+        - connectionName
         - sourceId
         - destinationId
         - status
       properties:
         name:
           type: string
           description: Optional name of the connection
+        connectionName:
```
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:1606]: Why not just called it prefix? tableNamePrefix? The connection already has a name. You're adding a field to allow prefixing right? Maybe I'm missing some bigger picture here, but it seems like you could use a really specific clean name here instead of something broad.

Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml:1606]: in the short term, it's a prefix but in one or two weeks, when namespace fields are available, this field becomes the actual namespace name for the destination.

On top of that destination namespace name, you can have an optional suffix namespace whenever the source supports it (in JDBC source connectors) but in all other cases, the suffix will be null.

To conclude, the connectionName can't be empty when used with non-JDBC sources.

Comment (michel-tricot) [reply airbyte-api/src/main/openapi/config.yaml:1606]: IMO: This is very complex & we should call things by their names.

if it is a prefix we should call it prefix. 
if it is a default namespace for record which doesn't have a namespace, then we should call it defaultNamespace

ConnectionName seems like a dead field at that point.

Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml]: oops i just realized I called it `namespaceDefault` instead of `defaultNamespace`... I hope that's fine...

File: airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java [hunk 2]
```diff
@@ -29,7 +29,8 @@
   public static final String TAP_CONFIG_JSON_FILENAME = "tap_config.json";
   public static final String TARGET_CONFIG_JSON_FILENAME = "target_config.json";
 
-  public static final String CATALOG_JSON_FILENAME = "catalog.json";
+  public static final String TAP_CATALOG_JSON_FILENAME = "tap_catalog.json";
```
Comment (ChristopheDuong) [inline airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java:32]: should we rename these constants to 
- source_catalog.json / destination_catalog.json?
- source_config.json / destination_config.json?

**Comment by ChristopheDuong** [comment]
I'm planning to add a test class for the `airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java` class that I introduced here but it should be ready for review in the mean time

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/Mapper.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import io.airbyte.commons.functional.CheckedConsumer;
+import io.airbyte.config.StandardSyncInput;
+import io.airbyte.config.StandardTargetConfig;
+
+public interface Mapper<T> extends CheckedConsumer<T, Exception>, AutoCloseable {
+
+  void start(StandardTargetConfig targetConfig, StandardSyncInput syncInput) throws Exception;
+
+  @Override
+  void accept(T message) throws Exception;
```
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/Mapper.java:36]: I think this should be a function instead of a consumer. We probably don't want to do in place transformation but instead copy and return the modified copy

File: airbyte-api/src/main/openapi/config.yaml [hunk 4]
```diff
@@ -1595,13 +1595,17 @@ components:
     ConnectionCreate:
       type: object
       required:
+        - namespaceDefault
```
Comment (michel-tricot) [inline airbyte-api/src/main/openapi/config.yaml:1598]: It is more correct if it is called `defaultNamespace`

File: airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java (hunk 2 above)
Comment (michel-tricot) [reply airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java:32]: Sounds good to me!

Comment (michel-tricot) [reply airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java:32]: If you do so, also rename the const names

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/AirbyteMapper.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.workers.protocols.airbyte;
+
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.workers.protocols.Mapper;
+
+public interface AirbyteMapper extends Mapper<AirbyteMessage> {
```
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/AirbyteMapper.java:30]: Why do you need this class?

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteDestination.java [hunk 6]
```diff
@@ -62,10 +62,11 @@ public void start(StandardTargetConfig targetConfig, Path jobRoot) throws IOExce
     Preconditions.checkState(targetProcess == null);
 
     IOs.writeFile(jobRoot, WorkerConstants.TARGET_CONFIG_JSON_FILENAME, Jsons.serialize(targetConfig.getDestinationConnectionConfiguration()));
-    IOs.writeFile(jobRoot, WorkerConstants.CATALOG_JSON_FILENAME, Jsons.serialize(targetConfig.getCatalog()));
+    IOs.writeFile(jobRoot, WorkerConstants.TARGET_CATALOG_JSON_FILENAME, Jsons.serialize(targetConfig.getCatalog()));
 
     LOGGER.info("Running target...");
-    targetProcess = integrationLauncher.write(jobRoot, WorkerConstants.TARGET_CONFIG_JSON_FILENAME, WorkerConstants.CATALOG_JSON_FILENAME).start();
+    targetProcess =
```
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteDestination.java:68]: NIT:
```
targetProcess =  integrationLauncher.write(
    jobRoot, 
    WorkerConstants.TARGET_CONFIG_JSON_FILENAME, 
    WorkerConstants.TARGET_CATALOG_JSON_FILENAME).start();


```

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+public class DefaultAirbyteMapper implements AirbyteMapper {
+
+  private StandardSyncInput syncInput;
+
+  public DefaultAirbyteMapper() {}
+
+  @Override
+  public void start(StandardTargetConfig targetConfig, StandardSyncInput syncInput) throws IOException, WorkerException {
+    this.syncInput = syncInput;
+    targetConfig.setCatalog(transformCatalog(targetConfig.getCatalog(), syncInput));
```
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:53]: This is a very bad practice to do inplace transformation.

Maybe the mapper needs 2 methods:
- one to transform the catalog
- one to transform the message

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  @Override
+  public void accept(AirbyteMessage message) throws IOException {
+    Preconditions.checkState(syncInput != null);
```
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:58]: This should be done in the constructor

Comment (michel-tricot) [reply airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:58]: "Fail fast"

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+import io.airbyte.workers.WorkerException;
+import java.io.IOException;
+
+/**
+ * We apply some transformations on the fly on the catalog (same should be done on records too) from
+ * the source before it reaches the destination. One of the transformation is to define the
+ * destination namespace where data will be stored and how to mirror (or not) the namespace used in
+ * the source (if any). This is configured in the UI through the syncInput.
+ */
+public class DefaultAirbyteMapper implements AirbyteMapper {
```
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:44]: This is not a default Airbyte mapper. It is a NamespacingMapper (or smth like that)

Comment (michel-tricot) [reply airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:44]: If you do so then you don't need to pass the whole syncInput, but just the defaultNamespace

Comment (michel-tricot) [reply airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:44]: "Only pass what you need"

**Review by michel-tricot — APPROVED** [review]
% suggested changes

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java (hunk 8 above)
Comment (ChristopheDuong) [reply airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:58]: But then I don't think i can pass it to the DefaultSyncWorker constructor?

it needs to be built within the `run()` method... but I guess if it's fine like that, then sure

Comment (michel-tricot) [reply airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteMapper.java:58]: what do you mean? 

why can you do `Preconditions.checkState(syncInput != null);` in accept but not in the constructor?

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  @Override
+  public StandardTargetConfig apply(final StandardTargetConfig inputConfig) {
+    final StandardTargetConfig config = Jsons.clone(inputConfig);
+    config.getCatalog().getStreams().forEach(s -> mutateStream(s.getStream(), defaultNamespace));
+    return config;
+  }
+
+  @Override
+  public AirbyteMessage apply(final AirbyteMessage inputMessage) {
+    final AirbyteMessage message = Jsons.clone(inputMessage);
```
Comment (ChristopheDuong) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java:56]: Is the clone operation really needed here? (but then we would mutate the `inputMessage` as a side-effect

In terms of performances, it would be avoiding a copy since the original message is not used anyway?
only the transformed one is passed to the destination

File: airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java (hunk 2 above)
Comment (cgardens) [reply airbyte-workers/src/main/java/io/airbyte/workers/WorkerConstants.java:32]: agreed

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java (hunk 10 above)
Comment (cgardens) [reply airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java:56]: probably doing the clone is safest for now to avoid any confusion. especially if we end up chaining these, i think it would confusing if they are mutating in place. but you're right there is some performance cost.

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    if (message.getCatalog() != null) {
+      message.getCatalog().getStreams().forEach(s -> mutateStream(s, defaultNamespace));
+    }
+    if (message.getRecord() != null) {
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java:60]: I think you can do this instead, which is a bit more readable.

```suggestion
    if (message.getType() == Type.RECORD) {
```

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  public StandardTargetConfig apply(final StandardTargetConfig inputConfig) {
+    final StandardTargetConfig config = Jsons.clone(inputConfig);
+    config.getCatalog().getStreams().forEach(s -> mutateStream(s.getStream(), defaultNamespace));
+    return config;
+  }
+
+  @Override
+  public AirbyteMessage apply(final AirbyteMessage inputMessage) {
+    final AirbyteMessage message = Jsons.clone(inputMessage);
+    if (message.getCatalog() != null) {
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java:57]: when would the sync worker ever be consuming catalog messages? unless there's a specific case i think we can remove this conditional.

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/Mapper.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+package io.airbyte.workers.protocols;
+
+import io.airbyte.config.StandardTargetConfig;
+
+public interface Mapper<T> {
+
+  StandardTargetConfig apply(StandardTargetConfig targetConfig);
+
+  T apply(T message);
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/Mapper.java:33]: can we have more explicit names for these methods. having them both be the same and just called apply is hard to parse. maybe `mapCatalog` and `mapMessage`?

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/Mapper.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+ * SOFTWARE.
+ */
+
+package io.airbyte.workers.protocols;
+
+import io.airbyte.config.StandardTargetConfig;
+
+public interface Mapper<T> {
+
+  StandardTargetConfig apply(StandardTargetConfig targetConfig);
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/Mapper.java:31]: why is this taking in `StandardTargetConfig` and opposed to `ConfiguredAirbyteCatalog`? The point is to map one catalog to another, right? `StandardTargetConfig` doesn't really have anything to do with it; it just happens to be a containing object.

File: airbyte-api/src/main/openapi/config.yaml (hunk 1 above)
Comment (cgardens) [reply airbyte-api/src/main/openapi/config.yaml]: > in the short term, it's a prefix but in one or two weeks, when namespace fields are available, this field becomes the actual namespace name for the destination.

in that second step, how will this work with not database sources that declare namespaces? if this a default and gets overridden by the namespace set by the source, won't tables in the destination still collide? e.g. if you were pulling from 2 stripe sources both will try to write to `stripe.users`?

File: airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+ * the source before it reaches the destination. One of the transformation is to define the
+ * destination namespace where data will be stored and how to mirror (or not) the namespace used in
+ * the source (if any). This is configured in the UI through the syncInput.
+ */
+public class NamespacingMapper implements Mapper<AirbyteMessage> {
+
+  private final String defaultNamespace;
+
+  public NamespacingMapper(String defaultNamespace) {
+    this.defaultNamespace = defaultNamespace;
```
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java:44]: what happens if the default namespace is null?

File: airbyte-api/src/main/openapi/config.yaml (hunk 1 above)
Comment (cgardens) [reply airbyte-api/src/main/openapi/config.yaml]: i'm still a little unsure if this a namespace default or namespace override. I'm a little fuzzy on what's going to happen if this value is not set. it looks like below we are going throw an NPE or end up with tables that are `null_users`. maybe i'm missing something?

**Review by cgardens — APPROVED** [review]
nice! some questions around readability but i think they are all straight forward. some bigger questions in the existing thread [here](https://github.com/airbytehq/airbyte/pull/2298#discussion_r587848536)

File: airbyte-api/src/main/openapi/config.yaml (hunk 1 above)
Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml]: Normally, the default namespace should always be defined to specify where data will end up in the destination.

The only case where it would be acceptable for it to be `null` is when the stream already has a namespace defined by the source (thus requires a namespace field introduced by #2228) AND when the user wants to mirror exactly the same namespace structure in the destination as in the source (replicate `public.users` from source in `public.users` in destination)

But for the short term, since destinations are still set with one (and it is required), it is not currently mandatory here immediately

Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml]: > in that second step, how will this work with non database sources that declare namespaces?

let's move that thread discussion over there where I will detail more what we do in the next steps.

https://github.com/airbytehq/airbyte/issues/2324

**Comment by ChristopheDuong** [comment]
I also made a short description in video for the changelog here https://www.loom.com/share/cc3cca41c37549779381abd847f2a003
